server/src/main/java/org/elasticsearch/search/query/QueryPhase.java

```diff
@@ -469,5 +474,90 @@ private static boolean canEarlyTerminate(IndexReader reader, SortAndFormats sort
         return true;
     }
 
+    /**
+    * Returns true if more than 50% of data in the index have the same value
+    * The evaluation is approximation based on finding the median value and estimating its count
+    * Returns true if most of the segments have duplicate data, false - otherwise
+    */
+    static boolean indexFieldHasDuplicateData(IndexReader reader, String field) throws IOException {
+        int duplicateSegments = 0;
+        int noDuplicateSegments = 0;
+        for (LeafReaderContext lrc : reader.leaves()) {
+            PointValues pointValues = lrc.reader().getPointValues(field);
+            int thresholdDocCount = pointValues.getDocCount()/2;
+            byte[] minValueAsBytes = pointValues.getMinPackedValue();
+            byte[] maxValueAsBytes = pointValues.getMaxPackedValue();
+            long minValue = LongPoint.decodeDimension(minValueAsBytes, 0);
+            long maxValue = LongPoint.decodeDimension(maxValueAsBytes, 0);
+            long medianCount = estimateMedianCount(pointValues, minValue, maxValue, thresholdDocCount);
+            if (medianCount > thresholdDocCount) {
+                duplicateSegments++;
+            } else {
+                noDuplicateSegments++;
+            }
+        }
+        return (duplicateSegments >= noDuplicateSegments);
```

Comment:
Can we compute the global count of all medians and compare with the total doc count divided by 2 ? Currently the heuristic does not take the size of segments into account.